kythe/cxx/indexer/proto/file_descriptor_walker.cc

```diff
@@ -297,6 +301,81 @@ void FileDescriptorWalker::VisitImports() {
   }
 }
 
+namespace {
+std::string SignAnnotation(
+    google::protobuf::GeneratedCodeInfo::Annotation annotation) {
+  std::string signature;
+  std::stringstream sig(signature);
+  bool first_node = true;
+  for (const auto& node : annotation.path()) {
+    sig << (first_node ? "" : ".") << node;
+    first_node = false;
+  }
+  return sig.str();
+}
+
+VName VNameForAnnotation(
+    VName context_vname,
+    google::protobuf::GeneratedCodeInfo::Annotation annotation) {
+  VName out;
+  out.set_corpus(context_vname.corpus());
+  out.set_path(annotation.source_file());
+  out.set_signature(SignAnnotation(annotation));
+  out.set_language(kLanguageName);
+  return out;
+}
+}  // anonymous namespace
+
+void FileDescriptorWalker::VisitGeneratedProtoInfo() {
+  if (!file_descriptor_->options().HasExtension(proto::generated_proto_info)) {
+    return;
+  }
+  proto::VName context_vname(file_name_);
+  GeneratedProtoInfo proto_info =
+      file_descriptor_->options().GetExtension(proto::generated_proto_info);
+  google::protobuf::GeneratedCodeInfo info = proto_info.generated_code_info();
```

Comment:
nit: save the unnecessary local and copy:

const google::protobuf::GeneratedCodeInfo& info = file_descriptor_->options().GetExtension(proto::generated_proto_info).generated_code_info();